circuit_passes/src/passes/loop_unroll/body_extractor.rs

```diff
@@ -74,36 +77,55 @@ impl SubcmpSignalCompare {
     }
 }
 
+#[derive(Debug, Eq, PartialEq)]
 struct ExtraArgsResult {
-    // NOTE: This collection and several intermediate collections that are used to build this
-    // one must use IndexMap/IndexSet to preserve insertion order to stabilize lit test output.
-    bucket_to_itr_to_ref: HashMap<BucketId, Vec<Option<(AddressType, AddressOffset)>>>,
-    bucket_to_args: IndexMap<BucketId, ArgIndex>,
+    // NOTE: These collections and several intermediate collections that are used to build them
+    // must use IndexMap/IndexSet to preserve insertion order to stabilize lit test output.
+    //
+    /// Table structure indexed first by load/store/call BucketId, then by iteration number
+    /// (i.e. the Vec index), containing the original memory references to use as arguments
+    /// when calling the extracted body function.
+    loc_to_itr_to_ref: IndexMap<BucketId, Vec<Option<(AddressType, AddressOffset)>>>,
+    loc_to_args: IndexMap<BucketId, ArgIndex>,
     num_args: usize,
 }
 
+impl PartialOrd for ExtraArgsResult {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        PartialOrd::partial_cmp(&self.num_args, &other.num_args)
+            .and_then(|_| self.loc_to_args.iter().partial_cmp(other.loc_to_args.iter()))
+            .and_then(|_| self.loc_to_itr_to_ref.iter().partial_cmp(other.loc_to_itr_to_ref.iter()))
```

Comment:
I could be wrong because I'm not super familiar with this, but I don't think this will end up computing the same ordering as the Ord trait below due to how `and_then` works. `and_then` should chain when the prior `Option` is not `None`, but could be `Some(Less)` or `Some(Greater)`, whereas `then_with` below chains only when the prior `Ordering` returned is `Equal`. So, I think here, you could have the following:
```rust
PartialOrd::partial_cmp(&self.num_args, &other.num_args) // returns Some(Less)
            .and_then(|_| self.loc_to_args.iter().partial_cmp(other.loc_to_args.iter())) // returns Some(Greater)
            .and_then(|_| self.loc_to_itr_to_ref.iter().partial_cmp(other.loc_to_itr_to_ref.iter())) // returns Some(Equal)
```

And your overall result will be `Equal`, but below, you'd have:
```rust
self.num_args
            .cmp(&other.num_args) // Returns Less
            .then_with(|| self.loc_to_args.iter().cmp(other.loc_to_args.iter())) // short-circuited
            .then_with(|| self.loc_to_itr_to_ref.iter().cmp(other.loc_to_itr_to_ref.iter())) // short-circuited
```
And thus returns `Less`.

So maybe we can use the ignored parameters in the `and_then` lambdas so we'll get the same behavior in both cases.

Like I said, could be wrong, but either way, I found the comparison a little hard to follow.

Reply:
@iangneal fixed. Since I had to implement Ord anyway, we can just use it.